core/src/smartcontracts/isi/asset.rs

```diff
@@ -275,46 +271,49 @@ pub mod isi {
         }
     }
 
+    /// Assert that the two assets have the same asset `definition_id`.
+    fn assert_matching_definitions<W: WorldTrait>(
+        source: &<Asset as Identifiable>::Id,
+        destination: &<Asset as Identifiable>::Id,
+        wsv: &WorldStateView<W>,
+        value_type: AssetValueType,
+    ) -> Result<(), Error> {
```

Comment:
Am I the only one who thinks about assertion functions like it should panic instead of returning `Result`? I'd prefer `check_` naming but that's a personal taste

Reply:
No a panic in this case doesn't signal the Iroha programmer's error it signals that the person who requested the ISI didn't consider the state of the blockchain. Since panics are not forwarded to clients as error messages, these `assert_` methods need to return a `Result`. 

`check_` and `assert_` have [different connotations](https://artoftesting.com/difference-between-assert-and-verify). 

> We use Assert when we have to validate critical functionality. If the feature fails then this makes the execution of further statements irrelevant. Hence, the test method is aborted as soon as failure occurs.

We cannot handle a `tx` that has the wrong type thinking it's the right type, we have no choice but to abort the execution. Same with `source` and `destination`. 

A check is usually something that can be recovered from: i.e. a `None` that can be replaced with `Some(Default::default())`.

Reply:
Makes sense